### PR TITLE
Assorted bag'o tweaks & fixes

### LINKED ISSUE #8039: Autoshutdown not working on a Koob Touch
Hi there,
as the title states I can't seem to get auto shutdown to work. Auto suspend on the other hand does work.

I'm using Kobo Touch (2gb). I have auto suspend set to 15min and auto shutdown set to 1 hr. I've tried a few different times for the auto shutdown, such as 0.5 hr, 0.2 hr and it wouldn't shut down itself. I understand it's a pretty minor issue but I figured I would post about it in case nobody was aware of the problem. Thanks and let me know if there is anything I can do to help!

### LINKED ISSUE #9567: Koreader freezes when pressing f2
* KOReader version: 2022-08
* Device: Arch linux laptop, x86_64

#### Issue
Koreader freezes when pressing f2 and cant close it with my wm shortcuts or buttons, im using awesomewm as my wm
tried with kill and doesnt close neither
#### Steps to reproduce
open Koreader
press f2
cry because you need to open the terminal to close it
##### `crash.log` (if applicable)
doesn't seem to create a crash log, its like frozen 

Android logs are kept in memory. Please go to Help → Bug Report to save these logs to a file.

Please try to include the relevant sections in your issue description.
You can upload the whole `crash.log` file on GitHub by dragging and
dropping it onto this textbox.

If you instead opt to inline it, please do so behind a spoiler tag:
<details>
  <summary>crash.log</summary>

```
<Paste crash.log content here>
```
</details>

### LINKED ISSUE #9573: Migration to new defaults config lost STARDICT_DATA_DIR
I had 'STARDICT_DATA_DIR' defined in defaults.persistent.lua to use the same stardict files directory in Koreader and in ColorDict. With migration to the new config system it was moved to defaults.legacy.lua and stopped working. I added it manually to the new 'defaults.system.lua' file and now it works.

### PR DESCRIPTION
**Pull request by NiLuJe**
This... started much smaller than it ended up being, but conflicting changes in unrelated or semi-related stuff made working and testing this all at the same time much saner and easier, with the drawback being a giant squashed commit with a lot of random crap in it :/.

----

* UIManager: Support more specialized update modes for corner-cases:
  * A2, which we'll use for the VirtualKeyboards keys (they'd... inadvertently switched to UI with the highlight refactor).
  * NO_MERGE variants of ui & partial (for sunxi). Use `[ui]` in ReaderHighlight's popup, because of a Sage kernel bug that could otherwise make it translucent, sometimes completely so (*sigh*).
* UIManager: Assorted code cleanups & simplifications.
* Logger & dbg: Unify logging style, and code cleanups.
* SDL: Unbreak suspend/resume outside of the emulator (fix #9567).
* NetworkMgr: Cache the network status, and allow it to be queried. (Used by AutoSuspend to avoid repeatedly poking the system when computing the standby schedule delay).
* OneTimeMigration: Don't forget about `NETWORK_PROXY` & `STARDICT_DATA_DIR` when migrating `defaults.persistent.lua` (fix #9573)
* WakeupMgr: Workaround an apparent limitation of the RTC found on i.MX5 Kobo devices, where setting a wakealarm further than UINT16_MAX seconds in the future would apparently overflow and wraparound... (fix #8039, many thanks to @yfede for the extensive deep-dive and for actually accurately pinpointing the issue!).
* Kobo: Handle standby transitions at full CPU clock speeds, in order to limit the latency hit.
* UIManager: Properly quit on reboot & exit. This ensures our exit code is preserved, as we exit on our own terms (instead of being killed by the init system). This is important on platforms where exit codes are semantically meaningful (e.g., Kobo).
* UIManager: Speaking of reboot & exit, make sure the Screensaver shows in all circumstances (e.g., autoshutdown, re: #9542)), and that there aren't any extraneous refreshes triggered. (Additionally, fix a minor regression since #9448 about tracking this very transient state on Kobo & Cervantes).
* Kindle: ID the upcoming Scribe.
* Bump base (https://github.com/koreader/koreader-base/pull/1524)

Requires https://github.com/koreader/koreader-base/pull/1524

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9569)

### CODE REVIEW

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9569)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @NiLuJe)

File: frontend/ui/uimanager.lua [hunk 1]
```diff
@@ -1134,26 +1157,27 @@ function UIManager:_repaint()
...
-        local widget = self._window_stack[i]
+        local window = self._window_stack[i]
+        local widget = window.widget
         -- paint if current widget or any widget underneath is dirty
-        if dirty or self._dirty[widget.widget] then
+        if dirty or self._dirty[widget] then
             -- pass hint to widget that we got when setting widget dirty
             -- the widget can use this to decide which parts should be refreshed
-            logger.dbg("painting widget:", widget.widget.name or widget.widget.id or tostring(widget))
+            logger.dbg("painting widget:", widget.name or widget.id or tostring(widget))
```
Comment (NiLuJe) [inline frontend/ui/uimanager.lua:1166]: The confusion between `window` and `widget` led to a very old and stupid bug, where, for anonymous widgets, this was logging the reference of the `window`, while both `show` & `close` log the reference of the `widget` (which was, obviously, always the intended behavior).

**Comment by NiLuJe** [comment]
Oh, joy, the PowerCover continues to impress with its brokenness.

Was testing the standby tweaks on a Sage after the H2O, and wondering why it failed to enter standby (or, well, suspend): guess what, cilix was preventing suspend (... outside of the cover, of course).

Yaaaay. -_-".

Anyway, fixed after a reboot, and the change helps there, too (it still shaves a good 25% of the timing, but when that 8ms at worst, it's much less impressive than shaving 80ms off 300 ;p).

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 4 of 10 files at r2, 14 of 17 files at r3, 5 of 5 files at r4, 4 of 4 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9569)* status: all files reviewed, 1 unresolved discussion (waiting on @Frenzie, @pazos, and @poire-z)

**Comment by NiLuJe** [comment]
Whelp, nothing's blown up so far, so I'll probably merge this tonight, before I start piling even more stuff on top of it... ;o).

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 4 of 4 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9569)* status: all files reviewed, 1 unresolved discussion (waiting on @Frenzie, @pazos, and @poire-z)

File: frontend/device/kindle/device.lua [hunk 2]
```diff
@@ -1188,7 +1224,8 @@ local pw4_set = Set { "0PP", "0T1", "0T2", "0T3", "0T4", "0T5", "0T6",
                   "16Q", "16R", "16S", "16T", "16U", "16V" }
 local kt4_set = Set { "10L", "0WF", "0WG", "0WH", "0WJ", "0VB" }
 local koa3_set = Set { "11L", "0WQ", "0WP", "0WN", "0WM", "0WL" }
-local pw5_set = Set { "1LG", "1Q0", "1PX", "1VD", "219", "21A", "2BH", "2BJ" }
+local pw5_set = Set { "1LG", "1Q0", "1PX", "1VD", "219", "21A", "2BH", "2BJ", "2DK" }
+local scribe_set = Set { "22D", "25T", "23A", "2AQ", "2AP", "1XH", "22C" }
```
Comment (hius07) [inline frontend/device/kindle/device.lua:1228]: Just for my curiosity, how did you get all of them?

Comment (NiLuJe) [reply frontend/device/kindle/device.lua:1228]: Depends on the FW version, currently it's in `/etc/deviceTypes.conf` (a similar file has existed throughout the FW history, but it wasn't necessarily always this one).

There were two new PW5 ids in 5.14.2, one in 5.14.3, and the Scribe appeared in 5.15.1 (i.e., it'll probably ship on 5.15.0).

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 4 of 4 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9569)* status: all files reviewed, 1 unresolved discussion (waiting on @Frenzie, @pazos, and @poire-z)

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 4 of 4 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9569)* status: all files reviewed, 1 unresolved discussion (waiting on @Frenzie, @pazos, and @poire-z)

**Comment by NiLuJe** [comment]
Whelp, big red button time ;).

File: frontend/ui/uimanager.lua [hunk 3]
```diff
@@ -710,19 +701,16 @@ function UIManager:getSecondTopmostWidget()
...
-            return nil
-        end
+        local widget = self._window_stack[i].widget
 
-        if sec.widget.name then
-            if sec.widget.name ~= "VirtualKeyboard" then
-                return sec.widget.name
+        if widget.name then
+            if widget.name ~= "VirtualKeyboard" then
+                return widget.name
```
Comment (weijiuqiao) [inline frontend/ui/uimanager.lua:708]: Curious if the `name` property is special that itself is a widget or this might be a bug?

Comment (NiLuJe) [reply frontend/ui/uimanager.lua:708]: I'm... not sure I understand the question?

(Very few widgets have the `name` field set, ReaderUI, FileManager & VirtualKeyboard are the main contenders).

Comment (NiLuJe) [reply frontend/ui/uimanager.lua:708]: Oh, yeah, no, that's on purpose ;). If it returns a string it's the name, if it returns a ref it's a ref to the widget, it return nil if no match ;).

The intended use-case was comparing against a *known* widget, so an equality test against a string or a ref doesn't change a thing for the caller.

I can see that you and autodim use it to actually poke at the widget's internal fields, in which case, you should definitely type-check the result (or tweak it to return a name, widget tuple if you find it more convenient). ;).

Comment (NiLuJe) [reply frontend/ui/uimanager.lua:708]: If you need something to test, the frontlight widget is named, too, and I think it popping up above the vocabbuilder might be entirely pluasible?

Comment (weijiuqiao) [reply frontend/ui/uimanager.lua:708]: Thanks for the explanation :) I use it only when a wiki window is shown and I try to check if it’s from a dictionary window so to inherit the latter’s button tweaks. For now I don’t see it possible to have a wiki window above a nil widget, so it should be fine, I suppose :) (just checked strings can be dotted too to get a nil so won’t break the code)

Comment (NiLuJe) [reply frontend/ui/uimanager.lua:708]: Oh, right, every string gets a metatable that points to the string standard library, so field access will work (funky!).

**Comment by poire-z** [comment]
Mentionning this because git blame links the related line to this PR :)
On my emulator (with koreader master), I have gesture one finger swipe top left edge to suspend (not using it, must have added it for testing stuff, and triggered it by error), and when doing that gesture while in Reader:
```
./luajit: frontend/ui/uimanager.lua:1547: attempt to index field 'event_handlers' (a nil value)
stack traceback:
 frontend/ui/uimanager.lua:1547: in function 'handleEvent'
 frontend/ui/widget/container/widgetcontainer.lua:82: in function 'propagateEvent'
 frontend/ui/widget/container/widgetcontainer.lua:100: in function 'handleEvent'
 frontend/ui/uimanager.lua:861: in function 'sendEvent'
 frontend/dispatcher.lua:974: in function 'execute'
 plugins/gestures.koplugin/main.lua:1079: in function 'handler'
 frontend/ui/widget/container/inputcontainer.lua:253: in function 'handleEvent'
 frontend/ui/uimanager.lua:861: in function 'sendEvent'
 frontend/ui/uimanager.lua:51: in function '__default__'
 frontend/ui/uimanager.lua:1385: in function 'handleInputEvent'
 frontend/ui/uimanager.lua:1483: in function 'handleInput'
 frontend/ui/uimanager.lua:1530: in function 'run'
 ./reader.lua:302: in main chunk
 [C]: at 0x56203012f7a3
```

**Comment by NiLuJe** [comment]
That's... mysterious. Nothing ever `nil`s UIManager's `event_handlers`, and it's a table setup at `init` time (which is done at `require` time in UIManager's case. o_O.

**Comment by poire-z** [comment]
So, `self` inside this call to `UIManager:suspend()` is not UIManager, but.. DeviceListener.
Dispatcher indeed sends the `RequestSuspend` event, whose only handler is:
https://github.com/koreader/koreader/blob/9473c70dccf475089c12e07b320f610fd4914c0c/frontend/device/devicelistener.lua#L320

#9657 has been a bit too random at its optimizations :/ The previous code was the right way to do it: https://github.com/koreader/koreader/pull/9657/files#diff-16859aefa5d4dbe879ade176c19b3d5038938075d456f5703bda2f376a2d4086

**Pull request merged**